File: python/2_Add_Two_Numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def printList(self, x):
        while (x != None):
            print(x.val)
            x = x.next

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = tail = None
        flag = 0
        while l1 != None or l2 != None:
            if l1 != None and l2 != None:
                node = ListNode((l1.val + l2.val + flag) % 10)
                flag = int((l1.val + l2.val + flag)/10)
                l1 = l1.next
                l2 = l2.next
            elif l1 == None and l2 != None:
                node = ListNode((l2.val + flag) % 10)
                flag = int((l2.val + flag)/10)
                l2 = l2.next
            elif l1 != None and l2 == None:
                node = ListNode((l1.val + flag) % 10)
                flag = int((l1.val + flag)/10)
                l1 = l1.next

            if head == None:
                head = tail = node
            else:
                tail.next =  node
                tail = node

        if flag == 1:
            node = ListNode(1)
            tail.next =  node
            tail = node

        return head

lnode1 = ListNode(2)
lnode2 = ListNode(4)
lnode3 = ListNode(3)
lnode1.next = lnode2
lnode2.next = lnode3

rnode1 = ListNode(5)
rnode2 = ListNode(6)
rnode3 = ListNode(4)
rnode1.next = rnode2
rnode2.next = rnode3

sol = Solution()
head = sol.addTwoNumbers(lnode1, rnode1)
head.printList(head)

        

File: cpp/2_AddTwoNumbers.cpp
#include <iostream>
using namespace std;


  //Definition for singly-linked list.
  struct ListNode {
      int val;
      ListNode* next;
      ListNode(int x) : val(x), next(NULL) {}
  };
 
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int x = 0;
        int y = 0;
        int add = 0;
        int flag = 0;
        ListNode* head = NULL;
        ListNode* curr = NULL;
        while(l1 || l2 || flag){
            if(l1){
                x = l1->val; 
                l1 = l1->next;
            }
            else{
                x = 0;
            }
            
            if(l2){
                y = l2->val; 
                l2 = l2->next;
            }
            else{
                y = 0;
            }
            
            add = (x+y+flag)%10;
            flag = (x+y+flag)/10;
            ListNode* node = new ListNode(add);
            
            if(!head){
                head = node;
                curr = node;
            }
            else{
                curr->next = node;
                curr = node;
            }
        }
        
        return head;
    }
};

int main()
{
    ListNode node11(2);
    ListNode node12(4);
    ListNode node13(3);
    node11.next = &node12;
    node12.next = &node13;

    ListNode node21(5);
    ListNode node22(6);
    ListNode node23(4);
    node21.next = &node22;
    node22.next = &node23;

    Solution sol;
    ListNode* ans = sol.addTwoNumbers(&node11, &node21);

    while(ans)
    {
        cout<<ans->val<<" ";
        ans = ans->next;
    }

    cout<<endl;
}